stetho/src/main/java/com/facebook/stetho/server/http/LeakySocket.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) 2014-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+package com.facebook.stetho.server.http;
+
+import android.net.LocalSocket;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+/**
+ * Utility to allow upgrading a normal HTTP connection to a raw socket (for use with the WebSocket
+ * implementation).  This requires that we read the previously buffered data and stitch together
+ * a magic input stream
+ */
+public class LeakySocket {
+  private final LocalSocket mSocket;
+  private final LeakyBufferedInputStream mLeakyInput;
+
+  LeakySocket(LocalSocket socket, LeakyBufferedInputStream leakyInput) {
+    mSocket = socket;
+    mLeakyInput = leakyInput;
+  }
+
+  public InputStream getInput() throws IOException {
+    return new CompositeInputStream(new InputStream[] {
+        new ByteArrayInputStream(mLeakyInput.clearBuffer()),
+        mSocket.getInputStream()
+    });
```

Comment:
Two concerns here:
1) Do we need the clearBuffer() machinery?  It looks like the contents remaining in the stream we return via this machinery are exactly the same as if we just returned mLeakyInput here.
2) The remainder of the stream is unbuffered.  Is that intentional?